### DISCUSSION TITLE
Transmission speeds - GAP8/STM32

### DISCUSSION THREAD

**Discussion by finnfi** [bitcraze/discussions#105] [Q&A] [answered]
Hi,

I have a question regarding transfer speeds. 
The flow of the program: 
1. GAP8 asks STM32 for cf state (position only)
2. STM32 sends state to GAP8
3. GAP8 takes picture with HIMAX camera (greyscale, RAW)
4. GAP8 send picture to client pc with WIFI

The CPX protocol is used for every transmission. 
Timing (used xTaskGetTickCount) (I believe 1 tick is 1 ms, right?) :
- 1. and 2. uses ca. 100-400 ticks (combined)
- 3. uses ca. 50-100 ticks
- 4. uses ca. 10 ticks 

Python client (opencv-viewer.py) reports a FPS of around 2.5. 

Question:
Should this not be possible to do faster? Especially the transmission between gap8 and stm?

**Comment by evoggy** [comment, marked as answer]
Hi!

Yes, this should be optimized.

**1 and 2:** Yes, this should be faster. One way to speed this up is to just push the data from the STM32 (instead of pulling from the GAP8).  Another way is to use the GAP8<->STM32 UART and implement your own protocol there. I've also created an issue for the ESP32<->STM32 UART performance (https://github.com/bitcraze/crazyflie-firmware/issues/1065). Hopefully there will be improvements in the coming weeks.

**3:** I've seen the same figures, and it's too slow. The camera should be faster.

**4:** The queues in the GAP8 should fit one raw image. The idea is that the image is captured, queued very fast and then sent to the ESP32/WiFi while the next image is captured. This should be fast, but there's an issue we've seen where the WiFi will stall every now and then causing lowered frame rate (https://github.com/bitcraze/aideck-esp-firmware/issues/12)

For the FPS this is low, I normally get 5-10 FPS (although I think it should be higher still).

**Reply by finnfi** [reply]
Thank you for your fast response!
I will try to implement the pushing strategy for 1. and 2., and will wait for further fixes for the other issues!

**Reply by Rwik2000** [reply]
did you try out 1 and 2? How did it go?

If you solved the issue, can you tell me how did you do that ?
